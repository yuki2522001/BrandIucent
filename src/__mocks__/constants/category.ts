import { Categories } from "@common-types/category";
import url from "@assets/images/icons/beverages.png";

const CATEGORY: Categories = {
  id: "1234",
  name: "Beverages",
  images: {
    src: url,
    alt: "This is icon beverages",
  },
};

const CATEGORY_MOCKING_LIST: Categories[] = [
  {
    id: "1234",
    name: "Beverages",
    images: {
      src: "@assets/images/icons/beverages.png",
      alt: "This is icon beverages",
    },
  },
  {
    id: "12345",
    name: "Dairy",
    images: {
      src: "@assets/images/icons/dairy.png",
      alt: "This is icon dairy",
    },
  },
  {
    id: "12345",
    name: "Meat",
    images: {
      src: "@assets/images/icons/meat.png",
      alt: "This is icon meat",
    },
  },
  {
    id: "12345",
    name: "Spices & Herbs",
    images: {
      src: "@assets/images/icons/spices-herbs.png",
      alt: "This is icon spices-herbs",
    },
  },
  {
    id: "12345",
    name: "Stationery",
    images: {
      src: "@assets/images/icons/stationery.png",
      alt: "This is icon stationery",
    },
  },
  {
    id: "12345",
    name: "Costmetic",
    images: {
      src: "@assets/images/icons/costmetic.png",
      alt: "This is icon costmetic",
    },
  },
];

export { CATEGORY, CATEGORY_MOCKING_LIST };
