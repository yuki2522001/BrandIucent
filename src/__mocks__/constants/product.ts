import { Product } from "@common-types/product";
import iceCreamSundae from "@assets/images/ice-cream-sundae.png";
import beefSteak from "@assets/images/beef-steak.png";

const PRODUCT_MOCKING: Product = {
  productId: "16466545464646547",
  categoryId: "1651454657522",
  productName: "Ice Cream Sundae",
  productDescription:
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam molestie ultrices donec pulvinar quis est bibendum suspendisse viverra.",
  productQuantity: 5,
  originalPrice: {
    value: 20,
    currency: "$",
  },
  discountPrice: {
    value: 20,
    currency: "$",
  },
  offerQuantity: "3kg",
  offer: false,
  bestSelling: true,
  popular: false,
  images: {
    src: iceCreamSundae,
    alt: "This is ice-cream-sundae images",
  },
};
const PRODUCT_MOCKING_LIST: Product[] = [
  {
    productId: "16466545464646547",
    categoryId: "1651454657522",
    productName: "Ice Cream Sundae",
    productDescription:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam molestie ultrices donec pulvinar quis est bibendum suspendisse viverra.",
    productQuantity: 5,
    originalPrice: {
      value: 20,
      currency: "$",
    },
    discountPrice: {
      value: 20,
      currency: "$",
    },
    offerQuantity: "3kg",
    offer: false,
    bestSelling: true,
    popular: false,
    images: {
      src: iceCreamSundae,
      alt: "This is ice-cream-sundae images",
    },
  }
];

export { PRODUCT_MOCKING, PRODUCT_MOCKING_LIST };
