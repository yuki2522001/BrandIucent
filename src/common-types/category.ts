export interface Categories {
  id: string;
  name: string;
  images: {
    src: string;
    alt: string;
  };
}