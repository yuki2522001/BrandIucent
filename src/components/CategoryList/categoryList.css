.categories {
  padding: 120px 0 0 125px;
}

.categories__title {
  margin-bottom: 69px;
  font-weight: 600;
  font-size: 24px;
  line-height: 36px;
  text-transform: capitalize;
  color: var(--text-primary);
}

.categoryList__row {
  display: flex;
  margin-bottom: 246px;
}

.categoryList__card {
  display: flex;
  flex-direction: column;
  margin-right: 55px;
  background: var(--light);
  border-radius: 5px;
  width: 235px;
  box-shadow: -50px 50px 200px var(--box--light);
}

.categoryList__checkbox {
  margin: 0 0 23px 73px;
  font-weight: 400;
  font-size: 24px;
  line-height: 36px;
  color: var(--text-primary);
}