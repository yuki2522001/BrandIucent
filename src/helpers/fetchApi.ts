import axios from "axios";

// Get data form server
const getData = async <T>(url: string): Promise<T> => {
  const res = await axios.get<T>(url);
  return res.data;
};

export {
  getData,
}