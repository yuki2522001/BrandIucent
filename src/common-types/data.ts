import { Categories } from "./category";
import { Product } from "./product";

export interface DataContextProps {
  products: Product[];
  setProducts: Function;
  searchValue: [] | undefined;
  setSearchValue: Function;
  categories: Categories[];
  setCategories: Function;
}