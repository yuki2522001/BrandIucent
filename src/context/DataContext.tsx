import { createContext, useMemo, useState } from "react";
import { Categories } from "@common-types/category";
import { DataContextProps } from "@common-types/data";
import { Product } from "@common-types/product";

const DataContext = createContext<DataContextProps>({} as DataContextProps);
const DataProvider: React.FC<{ children: JSX.Element[] | JSX.Element }> = ({
  children,
}) => {
  const [searchValue, setSearchValue] = useState<[]>([]);
  const [categories, setCategories] = useState<Categories[]>([]);
  const [products, setProducts] = useState<Product[]>([]);

  const value = {
    products,
    setProducts,
    searchValue,
    setSearchValue,
    categories,
    setCategories,
  };

  return <DataContext.Provider value={value}>{children}</DataContext.Provider>;
};

export { DataProvider, DataContext };
