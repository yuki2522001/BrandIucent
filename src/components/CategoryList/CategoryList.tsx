import useSWR from "swr";
import React, { useContext, useEffect } from "react";
import { Categories } from "@common-types/category";
import { getData } from "@helpers/fetchApi";
import { CATEGORIES_URL } from "@constants/url";
import { DataContext } from "@context/DataContext";
import CategoryCard from "@components/CategoryCard/CategoryCard";
import CategoryCheck from "@components/CategoryCheck/CategoryCheck";
import "./categoryList.css";

interface CategoryProps {
  id?: string;
  type: string;
  isSelect?: boolean;
  isCheckbox?: boolean;
  onToggleCategory: Function;
  selectedCategories: string[];
}

const CategoryList: React.FC<CategoryProps> = ({
  id,
  type,
  isSelect,
  isCheckbox,
  onToggleCategory,
  selectedCategories
}) => {
  // fetch data with useSWR
  const { data } = useSWR(CATEGORIES_URL, getData<Categories[]>);
  const { setCategories, categories } = useContext(DataContext);

  useEffect(() => {
    setCategories(data);
  }, [data]);

  let className = "";
  switch (type) {
    case "row":
      className = "categoryList__row";
      break;
    case "column":
      className = "categoryList__column";
      break;
    default:
      break;
  }

  const renderCategoryList = () => {
    return categories?.map((category: Categories) => (
      <div key={category.id}>
        {isSelect && (
          <div className="categoryList__card">
            <CategoryCard
              onToggleCategory={onToggleCategory}
              category={category}
              categoryId={category.id}
            />
          </div>
        )}
        {isCheckbox && (
          <div className="categoryList__checkbox">
            <CategoryCheck
              onToggleCategory={onToggleCategory}
              categoryId={category.id}
              checked={category.id == id ? true : false}
              text={category.name}
              value={selectedCategories.includes(category.id)}
            />
          </div>
        )}
      </div>
    ));
  };

  return (
    <div data-testid="category-list" className={className}>
      {isCheckbox &&
        <div className="categoryList__checkbox">
          <CategoryCheck
            onToggleCategory={onToggleCategory}
            categoryId=""
            text="All"
            checked={id ? false : true}
            value={!selectedCategories.length}
          />
        </div>
      }
      {renderCategoryList()}
    </div>
  );
};

export default CategoryList;
