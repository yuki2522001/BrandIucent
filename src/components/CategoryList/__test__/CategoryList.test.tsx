import "@testing-library/jest-dom";
import CategoryList from "../CategoryList";
import mockAxios from "@__mocks__/axios";
import { useState } from "react";
import { fireEvent, render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import { DataContextProps } from "@common-types/data";
import { DataContext } from "@context/DataContext";
import { CATEGORY_MOCKING_LIST } from "@__mocks__/constants/category";
import { CATEGORIES_URL } from "@constants/url";
import { getData } from "@helpers/fetchApi";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

const contextProductMock: DataContextProps = {
  searchValue: [],
  setSearchValue: jest.fn(),
  categories: CATEGORY_MOCKING_LIST,
  setCategories: jest.fn(),
  products: [],
  setProducts: jest.fn(),
};

describe("CategoryList component", () => {
  beforeEach(() => {
    (useState as jest.Mock).mockImplementation(
      jest.requireActual("react").useState
    );
  });

  afterEach(() => {
    mockAxios.reset();
  });

  const setup = () => {
    const history = createMemoryHistory();
    const utils = render(
      <DataContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <CategoryList
            type="row"
            isSelect={true}
            onToggleCategory={() => {}}
            selectedCategories={[]}
          />
        </Router>
      </DataContext.Provider>
    );
    const input = utils.getByTestId("category-list") as HTMLInputElement;
    return {
      input,
      ...utils,
    };
  };

  test("get categories item should call", async () => {
    mockAxios.get.mockResolvedValueOnce({ data: CATEGORY_MOCKING_LIST });
    const result = await getData(CATEGORIES_URL);
    expect(mockAxios.get).toHaveBeenCalledWith(CATEGORIES_URL);
    expect(result).toEqual(CATEGORY_MOCKING_LIST);
  });

  test("should render categoryList - type column component", () => {
    const history = createMemoryHistory();
    const { getByTestId } = render(
      <DataContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <CategoryList
            type="column"
            isSelect={true}
            onToggleCategory={() => {}}
            selectedCategories={[]}
          />
        </Router>
      </DataContext.Provider>
    );
    expect(getByTestId("category-list")).toBeInTheDocument();
  });

  test("should render categoryList - type row component", () => {
    const history = createMemoryHistory();
    const { getByTestId } = render(
      <DataContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <CategoryList
            type="row"
            isSelect={true}
            onToggleCategory={() => {}}
            selectedCategories={[]}
          />
        </Router>
      </DataContext.Provider>
    );
    expect(getByTestId("category-list")).toBeInTheDocument();
  });

  test("should render product by search category", () => {
    const { input } = setup();
    fireEvent.change(input, { target: { id: "1651999177368" } });
    expect(input.id).toEqual("1651999177368");
  });

  test("matches snapshot", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <DataContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <CategoryList
            type="select"
            isSelect={true}
            onToggleCategory={() => {}}
            selectedCategories={[]}
          />
        </Router>
      </DataContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
