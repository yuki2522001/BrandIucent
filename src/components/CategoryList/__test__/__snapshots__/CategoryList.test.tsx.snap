// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryList component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class=""
    data-testid="category-list"
  >
    <div>
      <div
        class="categoryList__card"
      >
        <div
          class="categoryCard"
          data-testid="category-card"
        >
          <figure
            class="categoryCard__image"
          >
            <img
              alt="This is icon beverages"
              src="@assets/images/icons/beverages.png"
            />
          </figure>
          <a
            class="categoryCard__text"
            href="/products/category/1234"
          >
            <p
              data-testid="category-item"
            >
              Beverages
            </p>
          </a>
        </div>
      </div>
    </div>
    <div>
      <div
        class="categoryList__card"
      >
        <div
          class="categoryCard"
          data-testid="category-card"
        >
          <figure
            class="categoryCard__image"
          >
            <img
              alt="This is icon dairy"
              src="@assets/images/icons/dairy.png"
            />
          </figure>
          <a
            class="categoryCard__text"
            href="/products/category/12345"
          >
            <p
              data-testid="category-item"
            >
              Dairy
            </p>
          </a>
        </div>
      </div>
    </div>
    <div>
      <div
        class="categoryList__card"
      >
        <div
          class="categoryCard"
          data-testid="category-card"
        >
          <figure
            class="categoryCard__image"
          >
            <img
              alt="This is icon meat"
              src="@assets/images/icons/meat.png"
            />
          </figure>
          <a
            class="categoryCard__text"
            href="/products/category/12345"
          >
            <p
              data-testid="category-item"
            >
              Meat
            </p>
          </a>
        </div>
      </div>
    </div>
    <div>
      <div
        class="categoryList__card"
      >
        <div
          class="categoryCard"
          data-testid="category-card"
        >
          <figure
            class="categoryCard__image"
          >
            <img
              alt="This is icon spices-herbs"
              src="@assets/images/icons/spices-herbs.png"
            />
          </figure>
          <a
            class="categoryCard__text"
            href="/products/category/12345"
          >
            <p
              data-testid="category-item"
            >
              Spices & Herbs
            </p>
          </a>
        </div>
      </div>
    </div>
    <div>
      <div
        class="categoryList__card"
      >
        <div
          class="categoryCard"
          data-testid="category-card"
        >
          <figure
            class="categoryCard__image"
          >
            <img
              alt="This is icon stationery"
              src="@assets/images/icons/stationery.png"
            />
          </figure>
          <a
            class="categoryCard__text"
            href="/products/category/12345"
          >
            <p
              data-testid="category-item"
            >
              Stationery
            </p>
          </a>
        </div>
      </div>
    </div>
    <div>
      <div
        class="categoryList__card"
      >
        <div
          class="categoryCard"
          data-testid="category-card"
        >
          <figure
            class="categoryCard__image"
          >
            <img
              alt="This is icon costmetic"
              src="@assets/images/icons/costmetic.png"
            />
          </figure>
          <a
            class="categoryCard__text"
            href="/products/category/12345"
          >
            <p
              data-testid="category-item"
            >
              Costmetic
            </p>
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
